### PR TITLE
Callback improvements

### PR DESCRIPTION
**Pull request by adler-j**
* Add step in some places where it was obviously missing
* Update repr in some places
* Made `func` style arguments give a DeprecationWarning, instead recommend use of composition.

### CODE REVIEW

File: odl/solvers/util/callback.py [hunk 1]
```diff
@@ -213,21 +217,30 @@ def __init__(self, results=None, function=None):
...
         >>> norm_function = lambda x: x.norm()
-        >>> callback = CallbackStore(function=norm_function)
+        >>> callback = CallbackStore() * norm_function
         """
         self.results = [] if results is None else results
         self.function = function
+        if function is not None:
+            warnings.warn('`function` argument is deprecated and will be '
+                          'removed in a future release. Use composition '
+                          'instead.',
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: Perhaps add `(see Examples in the documentation)`

File: odl/solvers/util/callback.py [hunk 2]
```diff
@@ -385,39 +401,63 @@ class CallbackPrintTiming(SolverCallback):
 
     """Callback for printing the time elapsed since the previous iteration."""
 
-    def __init__(self):
-        """Initialize a new instance."""
+    def __init__(self, fmt='Time elapsed = {:<5.03f} s', step=1):
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: This is almost a bit too long for the signature. Consider using `None`.

File: odl/solvers/util/callback.py [hunk 3]
```diff
@@ -385,39 +401,63 @@ class CallbackPrintTiming(SolverCallback):
...
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        fmt : string, optional
+            Formating that should be applied.
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: While correct, "formating" with one "t" is less common.

Also print the usage pattern, like
```
Will be used as ::

    print(fmt.format(t_current - t_previous))
```

File: odl/solvers/util/callback.py [hunk 4]
```diff
@@ -196,8 +197,11 @@ def __init__(self, results=None, function=None):
             List in which to store the iterates.
             Default: new list (``[]``)
         function : callable, optional
+            Deprecated, use composition instead.
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: Point to `Examples` for users to see what you mean.

File: odl/solvers/util/callback.py [hunk 5]
```diff
@@ -385,39 +401,63 @@ class CallbackPrintTiming(SolverCallback):
...
         """Initialize a new instance.
 
         Parameters
         ----------
         func : callable, optional
+            Deprecated, use composition instead.
             Functional that should be called on the current iterate before
             printing. Default: print current iterate.
         fmt : string, optional
             Formating that should be applied. Default: print representation.
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: Remove `Default: ...`

State usage pattern.

File: odl/solvers/util/callback.py [hunk 6]
```diff
@@ -427,34 +467,49 @@ def __init__(self, func=None, fmt='{!r}'):
...
+        >>> callback = CallbackPrint(fmt='{0:.2f}') * np.sum
         >>> callback([1, 2])
         3.00
         """
-        self.fmt = str(fmt)
+        self.func = func
+        if func is not None:
+            warnings.warn('`func` argument is deprecated and will be removed '
+                          'in a future release. Use composition instead.',
+                          DeprecationWarning)
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: Same comment as above

File: odl/solvers/util/callback.py [hunk 7]
```diff
@@ -385,39 +401,63 @@ class CallbackPrintTiming(SolverCallback):
...
     """Callback for printing the current value."""
 
-    def __init__(self, func=None, fmt='{!r}'):
+    def __init__(self, func=None, fmt='{!r}', step=1):
         """Initialize a new instance.
 
         Parameters
         ----------
         func : callable, optional
+            Deprecated, use composition instead.
```
Comment (kohr-h) [inline odl/solvers/util/callback.py]: See above

File: odl/solvers/util/callback.py (hunk 2 above)
Comment (adler-j) [inline odl/solvers/util/callback.py]: Well `None` doesn't say anything, this is very clear in what it will do so that's why i changed it

**Comment by adler-j** [comment]
Merge after CI

**Pull request merged**